src/grammar_mutator.c

```diff
@@ -259,36 +261,73 @@ size_t afl_custom_trim(my_mutator_t *data, uint8_t **out_buf) {
 int32_t afl_custom_post_trim(my_mutator_t *data, int success) {
 
   if (success) {
+    // Keep track that we will need to rewrite the tree file later
+    data->trim_was_effective = true;
 
-    // Update the trimming step
-    size_t remaining_node_size = data->tree_cur->non_terminal_node_list->size;
-    size_t remaining_edge_size = data->tree_cur->recursion_edge_list->size;
-
-    // Update the corresponding tree file
-    write_tree_to_file(data->trimmed_tree, data->tree_fn_cur);
-
+    // Swap in the trimmed tree as our current tree:
     tree_free(data->tree_cur);
     data->tree_cur = data->trimmed_tree;
 
     // Update the non-terminal node list
     tree_get_non_terminal_nodes(data->tree_cur);
-    for (uint32_t i = 0; i < data->cur_subtree_trimming_step; ++i)
+
+    // The idea here is that the results that we've already tried are always at
+    // the start of the non_terminal_node_list, because whenever we get a "yes" result
+    // the nodes get trimmed out and they are always the last thing we've popped off the list.
+    //
+    // So to avoid repeating the previous attempts, just pop them out of the list right away:
+    for (uint32_t i = 0; i < data->finished_subtree_trimming_nodes; ++i)
       list_pop_front(data->tree_cur->non_terminal_node_list);
+
+    // We are now pointing *at* the node we just replaced.
+    // So we want to skip it and its children!
+    node_t * node = list_pop_front(data->tree_cur->non_terminal_node_list);
+    if (node) {
+
+      // Remember to skip these next time as well:
+      data->finished_subtree_trimming_nodes += 1 + node->non_term_size;
+
+      for (uint32_t i = 0; i < node->non_term_size; ++i)
+        list_pop_front(data->tree_cur->non_terminal_node_list);
+
+    }
+
     size_t new_node_size = data->tree_cur->non_terminal_node_list->size;
 
-    data->cur_subtree_trimming_step += (remaining_node_size - new_node_size);
+    // Set our progress equal to the remaining un-tested nodes in the list:
+    data->cur_subtree_trimming_step += data->total_subtree_trimming_steps - new_node_size;
```

Comment:
Shouldn't this be `data->cur_subtree_trimming_step = data->total_subtree_trimming_steps - new_node_size;` instead of using `+=`?

Reply:
Yep. Good catch!